File: elixir/max.exs
#
# Find the biggest element of a non-empty array
#

defmodule Solution do
  def max(list), do: Enum.reduce(list, fn(a, b) -> if a > b, do: a, else: b end)
end

defmodule Manual do
  def max([head|tail]), do: get_max(tail, head)
  defp get_max([], max), do: max

  defp get_max([head|tail], max) do
    cond do
      head > max -> get_max(tail, head)
      true -> get_max(tail, max)
    end
  end
end

IO.inspect(Solution.max([1, 5, 3, 10]))
IO.inspect(Manual.max([1, 5, 3, 10]))

File: javascript/max.js
/* Find the biggest element of a non-empty array */

class Solution {
    static max(arr) {
        return arr.reduce((a, b) => {
            return a > b ? a : b;
        });
    }

    static recursive(arr, m) {
        if (m === undefined)
            m = arr[0];

        if (arr.length === 0)
            return m;

        return this.recursive(arr.slice(1), arr[0] > m ? arr[0] : m);
    }
}

let list = [0, 24, 13, 101, 7, 13, 77];

console.log(Solution.max(list));
console.log(Solution.recursive(list));
